### PR TITLE
Add support for kicad 5 primitives (roundrect, polygon, custom pad)

### PR DESCRIPTION
**Pull request by poeschlr**
**Work in progress** Feedback welcome

ToDo:
- [x] Add pad type roundrect with its additional attribute round radius
   - [x] Add initializers and pad type to Pad node
   - [x] Add support in pad serializer
      - My kicad nightly version is a bit inconsistent with where it puts whitespace in the fileformat. It does not put a space between layer info and round rect radius ratio. (Ignored for now.)
- [x] Add support for graphical polygon
   - [x] Find out if it is feasible to refactor the polyline node such that it and the graphical polygon can share code?
       - It is possible. Is now implemented
   - [x] Implement node for graphical polygon
   - [x] Implement serializer
- [x] Add support for custom pads
   - [x] Add support in pad node
   - [x] Add support in serializer (re use serializer of other nodes)
      - [x] polygon
      - [x] line
      - [x] arc
      - [x] circle
- [ ] Advanced features
  - [ ] cut, combine polygons (The later is not really necessary, but could result in nicer looking files)
     - [x] smaller polygon fully within larger polygon 
        - [ ] raise exception if not
     - [ ] polygons only need to partly overlap
         - [ ] check if the resulting polygon is still a single polygon (fail otherwise)
  - [x] support specialized nodes
    - [x] PolygoneLine
    - [x] Filled Rect (Filled arcs and circles might also be a good idea to have.)
  - [ ] check that all primitives intersect each other ("orphaned" primitives are not allowed)
    - kicad can read files that have pads with orphaned primitives. But i assume there is a reason why the devs build in a function not to allow making such a pad(Maybe DRC can not deal with it)

### CODE REVIEW

**Comment by pointhi** [comment]
graphical polygons are already implemented: https://github.com/pointhi/kicad-footprint-generator/issues/117, but it's at the moment not easily possible to create cutouts (I would like to abstract the hack KiCad currently does)

**Comment by poeschlr** [comment]
Do you have any plans for this abstraction? I don't want to implement it in a way that does not fulfill your standards.

**Comment by poeschlr** [comment]
I now have a first implantation of a data container for both currently implemented Polygon like Nodes.
As i am currently not very creative i named it PolyPoints for now. I might rename it if i can think of a better name.

To keep backwards compatibility this class accepts both the point list parameter of the Polygon and PolygoneLine Node (nodes and polygone)

This same container will then also be useful for the custom pad.

**Comment by poeschlr** [comment]
I have a first implementation for the custom pad. For now only  the polygon primitive is supported. (For others i need to do some testing to find out how they work.)

**Comment by pointhi** [comment]
> Do you have any plans for this abstraction? I don't want to implement it in a way that does not fulfill your standards.

Well, my only thought about this at the moment would be that a Polygone can be given an abritrary number of cutouts and this is internally converted to a single polyline as it's done in "Bitmap-to-component".

Allowing only one cutout for now should be enough for most use-cases (like capacitor-footprints) and this should be achivable without much hassle (search the closest points between inner and outer layer -> duplicate points and connect them together). Allowing multiple cutouts would require more complex algorithm where I don't know currently one.

**Comment by poeschlr** [comment]
You can take a look at what i have so far. Everything except the more advanced features is now done.

**Comment by pointhi** [comment]
The code changes look good so far. Only some of the style issues need to to be fixed.

:+1:

File: KicadModTree/PolygonPoints.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        self._initNodes(**kwargs)
+
+    def _initNodes(self, **kwargs):
+        self.nodes = []
+        if 'nodes' in kwargs:
+            for n in kwargs['nodes']:
+                self.nodes.append(Point2D(n))
+            if 'polygone' in kwargs:
+                raise KeyError('Use of "nodes" and "polygone" parameter at the same time is not supported.')
+        elif 'polygone' in kwargs:
```
Comment (pointhi) [inline KicadModTree/PolygonPoints.py]: Add deprecation warning:

```python
warnings.warn(
            "polygone argument is deprecated, use nodes instead",
            DeprecationWarning
)
```

Comment (poeschlr) [reply KicadModTree/PolygonPoints.py]: Good idea. Added. (Will be fun running one of the larger generator scripts. There will be a lot of warnings displayed.)

**Comment by pointhi** [comment]
I would ignore this for now. This is not a user application and we can define constraints like "cutout is not allowed to overlap with the outline or other cutouts".

**Comment by poeschlr** [comment]
I have now implemented a simple cut feature. The more complex nodes are now also supported as primitives for the custom pad.
As i make no checks at all i added a warning to the function call. I think we can implement such checks later on. (And make the cut function more resilient)

**Comment by poeschlr** [comment]
@pointhi could you take another look at this? The functionality provided by this would help to create better qfn generators.

File: KicadModTree/Point.py [hunk 2]
```diff
@@ -136,6 +137,9 @@ def __repr__(self):
     def __str__(self):
         return "(x={x}, y={y})".format(**self.__dict__())
 
+    def distanceTo(self, other):
```
Comment (pointhi) [inline KicadModTree/Point.py]: Move up to ```round_to``` function, rename to ```distance_to``` for name concistency and use ```Point2D.__arithmetic_parse``` to get the value for ```other```

Comment (pointhi) [reply KicadModTree/Point.py]: And add a docstring

File: KicadModTree/nodes/base/Pad.py [hunk 3]
```diff
@@ -166,3 +216,6 @@ def _getRenderTreeText(self):
         render_text += '({})'.format(' '.join(render_strings))
 
         return render_text
+
+    def addPrimitive(self, p):
```
Comment (pointhi) [inline KicadModTree/nodes/base/Pad.py]: add docstring

File: KicadModTree/PolygonPoints.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+                    pi = i
+                    pj = j
+                    min_distance = d
+
+        return (pi, pj)
+
+    def getPoints(self):
+        return self.nodes
+
+    def cut(self, other):
```
Comment (pointhi) [inline KicadModTree/PolygonPoints.py]: add docstring

File: KicadModTree/PolygonPoints.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        for n in self.nodes:
+            yield n
+
+    def __getitem__(self, idx):
+        return self.nodes[idx]
+
+    def __len__(self):
+        return len(self.nodes)
+
+    def findNearestPoints(self, other):
```
Comment (pointhi) [inline KicadModTree/PolygonPoints.py]: add docstring

File: KicadModTree/nodes/base/Polygon.py [hunk 6]
```diff
@@ -76,3 +67,6 @@ def _getRenderTreeText(self):
         render_text += "]"
 
         return render_text
+
+    def cut(self, other):
```
Comment (pointhi) [inline KicadModTree/nodes/base/Polygon.py]: add docstring

File: KicadModTree/PolygonPoints.py [hunk 7]
```diff
@@ -108,9 +108,26 @@ def findNearestPoints(self, other):
...
     def cut(self, other):
+        r""" Cut other polygon points from self
+
+        As kicad has no native support for cuting one polygon from the other,
+        the cut is done by connecting the nearest points of the two polygons
+        with two lines on top of each other.
+
+        This function assumes that the other polygon is fully within this one.
+        It also assumes that connecting the two nearest points creates a valid
+        polygon. (There are no geomtry checks)
```
Comment (pointhi) [inline KicadModTree/PolygonPoints.py]: geomtry -> geometry

File: KicadModTree/Point.py [hunk 8]
```diff
@@ -76,6 +76,15 @@ def round_to(self, base):
         return Point2D({'x': round(self.x / base) * base,
                         'y': round(self.y / base) * base})
 
+    def distance_to(self, value):
+        r"""Distance between selfe and another point
```
Comment (pointhi) [inline KicadModTree/Point.py]: selfe -> this and another point

**Comment by pointhi** [comment]
Thanks for your hard work. Should be file now 👍

**Pull request merged**